src/NuGet.Core/NuGet.CommandLine.XPlat/Commands/PackageReferenceCommands/ListPackage/ListPackageCommandRunner.cs

```diff
@@ -224,6 +256,63 @@ public async Task ExecuteCommandAsync(ListPackageArgs listPackageArgs)
             GetVersionsFromDict(packages, packagesVersionsDict, listPackageArgs);
         }
 
+        /// <summary>
+        /// Fetches deprecation info for all of the packages found in a project.
+        /// </summary>
+        /// <param name="packages">The packages found in a project.</param>
+        /// <param name="listPackageArgs">List args for the token and source provider</param>
+        private async Task GetDeprecationInfoAsync(
+            IEnumerable<FrameworkPackages> packages,
+            ListPackageArgs listPackageArgs)
+        {
+            var providers = Repository.Provider.GetCoreV3();
+
+            // Handling concurrency and throttling variables
+            var maxTasks = Environment.ProcessorCount;
+            var contactSourcesRunningTasks = new List<Task>();
+            var currentVersionsRequests = new List<Task>();
+
+            // Unique dictionary for packages and list of versions to handle different sources
+            var packagesVersionsDict = new Dictionary<string, IList<IPackageSearchMetadata>>();
+            AddPackagesToDict(packages, packagesVersionsDict);
+
+            // Clone and filter package versions to avoid duplicate requests
+            // and to avoid collection being enumerated to be modified.
+            var distinctPackageVersionsDict = GetUniqueResolvedPackages(packages);
+
+            // Prepare requests for each of the packages
+            foreach (var packageIdAndVersions in distinctPackageVersionsDict)
+            {
+                foreach (var packageVersion in packageIdAndVersions.Value)
+                {
+                    currentVersionsRequests.AddRange(
+                        PrepareCurrentVersionsRequests(
+                            packageIdAndVersions.Key,
+                            packageVersion,
+                            listPackageArgs,
+                            providers,
+                            packagesVersionsDict));
+                }
+            }
+
+            // Make the calls to the sources
+            foreach (var currentVersionTask in currentVersionsRequests)
```

Comment:
This throttling can be made more generic.

Reply:
See https://github.com/NuGet/NuGet.Client/pull/2959/commits/122c990fb3fa8b8aeb13f79a848798498eed582a